File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.cpp
class Solution
{
public:
    int lengthOfLongestSubstring(string s)
    {
        int left = 0, ans = 0;
        unordered_map<char, int> m;
        for (int i = 0; i < s.size(); i++)
        {
            if (m.find(s[i]) != m.end())
            {
                left = max(m[s[i]] + 1, left);
            }
            m[s[i]] = i;
            ans = max(ans, i - left + 1);
        }
        return ans;
    }
};

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> mpp = new HashMap<Character, Integer>();
        
        int left = 0, len = 0;
        for (int right = 0; right < s.length(); right++)
        {
            if(mpp.containsKey(s.charAt(right)))
                left = Math.max(mpp.get(s.charAt(right)) + 1, left);
                
            mpp.put(s.charAt(right), right);
            len = Math.max(len, right - left + 1);
        }
        return len;
    }
}